### ISSUE TITLE
runUI allow other than Aff monad

### ISSUE THREAD

**Issue by BebeSparkelSparkel** [purescript-halogen/purescript-halogen#700]
https://github.com/purescript-halogen/purescript-halogen/blob/e6ac20be9531356d9c01fd8cab9e3ddf87c27c9a/src/Halogen/Aff/Driver.purs#L110-L115

I'm trying to use runUI in the StateT transformer. Is it possible to liftAff the `Component h f i o Aff`?

**Comment by JordanMartinez** [comment]
I'm pretty sure you can use `hoistAff` or something of that nature to do this.

However, using `StateT` as your underlying monad will produce problems. I believe there's another issue somewhere here that describes that further if you search for it. The workaround is to use a `ReaderT { state :: Ref s }` or something like that.

**Comment by garyb** [comment]
Yeah, `StateT` won't behave the way you want it to, as the state resets at the start of every computation when you weave it in using `hoist`. There's an old (closed I think) issue that talks about it a bit.

Realistically, any `m` you use has to be isomorphic to `ReaderT` over `Aff` as just about everything else will not behave the way we'd like. I looked into abstracting things to use class constraints rather than concrete `Aff`, so a more MTL-style of handling things could be done, but the constraints it imposed, requiring `Parallel`, `MonadFork`, etc. rule out most transformers anyway (there's no instance for these things for `StateT`). It'd also suffer a performance penalty with the extra overhead of dictionary allocations that would be needed throughout all of Halogen's machinery, and Halogen isn't going to win any benchmark contests as it is.

The `ReaderT` + `Ref` or `AVar` pattern is quite well used. It's not perfectly ergonomic since you have to use `Initializer`s to pull things out of there to populate the initial state of components and stuff. If you are using an `AVar` or `Ref` to mutate things over time you have to be a little careful, since you're essentially dealing with something like a threaded program if any of your component `eval`s are async, with the potential for race conditions and such that entails. This would all be true even if `State` worked anyway though.

**Comment by Mr-Andersen** [comment]
> rule out most transformers anyway

This *will* work for simple transformers over `Aff` that just have some more context with them (so, yeah, `ReaderT Aff`), and I'd argue that's quite common.

This limitation makes some cases unimplementable without major performance loss or some good ol `unsafeCoerce`. Maybe there are new developments that can change status of this issue?
